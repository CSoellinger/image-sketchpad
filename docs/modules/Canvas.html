<!DOCTYPE html><html class="default"><head><meta charSet="utf-8"/><meta http-equiv="x-ua-compatible" content="IE=edge"/><title>Canvas | image-sketchpad - v1.0.4</title><meta name="description" content="Documentation for image-sketchpad - v1.0.4"/><meta name="viewport" content="width=device-width, initial-scale=1"/><link rel="stylesheet" href="../assets/style.css"/><link rel="stylesheet" href="../assets/highlight.css"/><script async src="../assets/search.js" id="search-script"></script></head><body><script>document.body.classList.add(localStorage.getItem("tsd-theme") || "os")</script><header><div class="tsd-page-toolbar"><div class="container"><div class="table-wrap"><div class="table-cell" id="tsd-search" data-base=".."><div class="field"><label for="tsd-search-field" class="tsd-widget search no-caption">Search</label><input type="text" id="tsd-search-field"/></div><ul class="results"><li class="state loading">Preparing search index...</li><li class="state failure">The search index is not available</li></ul><a href="../index.html" class="title">image-sketchpad - v1.0.4</a></div><div class="table-cell" id="tsd-widgets"><div id="tsd-filter"><a href="#" class="tsd-widget options no-caption" data-toggle="options">Options</a><div class="tsd-filter-group"><div class="tsd-select" id="tsd-filter-visibility"><span class="tsd-select-label">All</span><ul class="tsd-select-list"><li data-value="public">Public</li><li data-value="protected">Public/Protected</li><li data-value="private" class="selected">All</li></ul></div> <input type="checkbox" id="tsd-filter-inherited" checked/><label class="tsd-widget" for="tsd-filter-inherited">Inherited</label></div></div><a href="#" class="tsd-widget menu no-caption" data-toggle="menu">Menu</a></div></div></div></div><div class="tsd-page-title"><div class="container"><ul class="tsd-breadcrumb"><li><a href="../modules.html">image-sketchpad - v1.0.4</a></li><li><a href="Canvas.html">Canvas</a></li></ul><h1>Module Canvas</h1></div></div></header><div class="container container-main"><div class="row"><div class="col-8 col-content"><section class="tsd-panel-group tsd-index-group"><h2>Index</h2><section class="tsd-panel tsd-index-panel"><div class="tsd-index-content"><section class="tsd-index-section "><h3>Classes</h3><ul class="tsd-index-list"><li class="tsd-kind-class tsd-parent-kind-module"><a href="../classes/Canvas.Canvas-1.html" class="tsd-kind-icon">Canvas</a></li></ul></section><section class="tsd-index-section "><h3>Type aliases</h3><ul class="tsd-index-list"><li class="tsd-kind-type-alias tsd-parent-kind-module"><a href="Canvas.html#Point" class="tsd-kind-icon">Point</a></li><li class="tsd-kind-type-alias tsd-parent-kind-module"><a href="Canvas.html#Stroke" class="tsd-kind-icon">Stroke</a></li></ul></section></div></section></section><section class="tsd-panel-group tsd-member-group "><h2>Type aliases</h2><section class="tsd-panel tsd-member tsd-kind-type-alias tsd-parent-kind-module"><a id="Point" class="tsd-anchor"></a><h3 class="tsd-anchor-link">Point<a href="#Point" aria-label="Permalink" class="tsd-anchor-icon"><svg xmlns="http://www.w3.org/2000/svg" class="icon icon-tabler icon-tabler-link" viewBox="0 0 24 24" stroke-width="2" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z" fill="none"></path><path d="M10 14a3.5 3.5 0 0 0 5 0l4 -4a3.5 3.5 0 0 0 -5 -5l-.5 .5"></path><path d="M14 10a3.5 3.5 0 0 0 -5 0l-4 4a3.5 3.5 0 0 0 5 5l.5 -.5"></path></svg></a></h3><div class="tsd-signature tsd-kind-icon">Point<span class="tsd-signature-symbol">:</span> <span class="tsd-signature-symbol">{ </span>x<span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">number</span><span class="tsd-signature-symbol">; </span>y<span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">number</span><span class="tsd-signature-symbol"> }</span></div><aside class="tsd-sources"><ul><li>Defined in <a href="https://github.com/CSoellinger/image-sketchpad/blob/main/src/Canvas/Draw.ts#L4">Canvas/Draw.ts:4</a></li></ul></aside><div class="tsd-comment tsd-typography"><div class="lead">
<p>Point - A point on the canvas set by coordinates.</p>
</div></div><div class="tsd-type-declaration"><h4>Type declaration</h4><ul class="tsd-parameters"><li class="tsd-parameter"><h5><span class="tsd-flag ts-flagReadonly">Readonly</span> x<span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">number</span></h5><div class="tsd-comment tsd-typography"><div class="lead">
<p>X coordinate of point.</p>
</div></div></li><li class="tsd-parameter"><h5><span class="tsd-flag ts-flagReadonly">Readonly</span> y<span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">number</span></h5><div class="tsd-comment tsd-typography"><div class="lead">
<p>Y coordinate of point.</p>
</div></div></li></ul></div></section><section class="tsd-panel tsd-member tsd-kind-type-alias tsd-parent-kind-module"><a id="Stroke" class="tsd-anchor"></a><h3 class="tsd-anchor-link">Stroke<a href="#Stroke" aria-label="Permalink" class="tsd-anchor-icon"><svg xmlns="http://www.w3.org/2000/svg" class="icon icon-tabler icon-tabler-link" viewBox="0 0 24 24" stroke-width="2" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z" fill="none"></path><path d="M10 14a3.5 3.5 0 0 0 5 0l4 -4a3.5 3.5 0 0 0 -5 -5l-.5 .5"></path><path d="M14 10a3.5 3.5 0 0 0 -5 0l-4 4a3.5 3.5 0 0 0 5 5l.5 -.5"></path></svg></a></h3><div class="tsd-signature tsd-kind-icon">Stroke<span class="tsd-signature-symbol">:</span> <span class="tsd-signature-symbol">{ </span>cap<span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">CanvasLineCap</span><span class="tsd-signature-symbol">; </span>color<span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">string</span><span class="tsd-signature-symbol">; </span>join<span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">CanvasLineJoin</span><span class="tsd-signature-symbol">; </span>maxWidth<span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">number</span><span class="tsd-signature-symbol">; </span>miterLimit<span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">number</span><span class="tsd-signature-symbol">; </span>points<span class="tsd-signature-symbol">: </span><a href="Canvas.html#Point" class="tsd-signature-type" data-tsd-kind="Type alias">Point</a><span class="tsd-signature-symbol">[]</span><span class="tsd-signature-symbol">; </span>width<span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">number</span><span class="tsd-signature-symbol"> }</span></div><aside class="tsd-sources"><ul><li>Defined in <a href="https://github.com/CSoellinger/image-sketchpad/blob/main/src/Canvas/Draw.ts#L19">Canvas/Draw.ts:19</a></li></ul></aside><div class="tsd-comment tsd-typography"><div class="lead">
<p>Stroke - A line between two or more <a href="Canvas.html#Point">Points</a>.</p>
</div></div><div class="tsd-type-declaration"><h4>Type declaration</h4><ul class="tsd-parameters"><li class="tsd-parameter"><h5><span class="tsd-flag ts-flagReadonly">Readonly</span> cap<span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">CanvasLineCap</span></h5><div class="tsd-comment tsd-typography"><div class="lead">
<p>{@link CanvasLineCap|Canvas line cap} of the stroke: &quot;butt&quot; | &quot;round&quot; | &quot;square&quot;.</p>
</div></div></li><li class="tsd-parameter"><h5><span class="tsd-flag ts-flagReadonly">Readonly</span> color<span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">string</span></h5><div class="tsd-comment tsd-typography"><div class="lead">
<p>Color of the stroke.</p>
</div></div></li><li class="tsd-parameter"><h5><span class="tsd-flag ts-flagReadonly">Readonly</span> join<span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">CanvasLineJoin</span></h5><div class="tsd-comment tsd-typography"><div class="lead">
<p>{@link CanvasLineJoin|Canvas line join} of the stroke: &quot;bevel&quot; | &quot;miter&quot; | &quot;round&quot;.</p>
</div></div></li><li class="tsd-parameter"><h5><span class="tsd-flag ts-flagReadonly">Readonly</span> max<wbr/>Width<span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">number</span></h5><div class="tsd-comment tsd-typography"><div class="lead">
<p>Max width of the stroke.</p>
</div></div></li><li class="tsd-parameter"><h5><span class="tsd-flag ts-flagReadonly">Readonly</span> miter<wbr/>Limit<span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">number</span></h5><div class="tsd-comment tsd-typography"><div class="lead">
<p>Miter limit of the stroke.</p>
</div></div></li><li class="tsd-parameter"><h5><span class="tsd-flag ts-flagReadonly">Readonly</span> points<span class="tsd-signature-symbol">: </span><a href="Canvas.html#Point" class="tsd-signature-type" data-tsd-kind="Type alias">Point</a><span class="tsd-signature-symbol">[]</span></h5><div class="tsd-comment tsd-typography"><div class="lead">
<p><a href="Canvas.html#Point">Points</a> of stroke.</p>
</div></div></li><li class="tsd-parameter"><h5><span class="tsd-flag ts-flagReadonly">Readonly</span> width<span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">number</span></h5><div class="tsd-comment tsd-typography"><div class="lead">
<p>Width of the stroke.</p>
</div></div></li></ul></div></section></section></div><div class="col-4 col-menu menu-sticky-wrap menu-highlight"><nav class="tsd-navigation primary"><ul><li class=""><a href="../modules.html">Modules</a></li><li class="current tsd-kind-module"><a href="Canvas.html">Canvas</a></li><li class=" tsd-kind-module"><a href="ImageSketchpad.html">Image<wbr/>Sketchpad</a></li><li class=" tsd-kind-module"><a href="index.html">index</a></li></ul></nav><nav class="tsd-navigation secondary menu-sticky"><ul><li class="tsd-kind-class tsd-parent-kind-module"><a href="../classes/Canvas.Canvas-1.html" class="tsd-kind-icon">Canvas</a></li><li class="tsd-kind-type-alias tsd-parent-kind-module"><a href="Canvas.html#Point" class="tsd-kind-icon">Point</a></li><li class="tsd-kind-type-alias tsd-parent-kind-module"><a href="Canvas.html#Stroke" class="tsd-kind-icon">Stroke</a></li></ul></nav></div></div></div><footer class="with-border-bottom"><div class="container"><h2>Legend</h2><div class="tsd-legend-group"><ul class="tsd-legend"><li class="tsd-kind-variable"><span class="tsd-kind-icon">Variable</span></li><li class="tsd-kind-function"><span class="tsd-kind-icon">Function</span></li><li class="tsd-kind-type-alias"><span class="tsd-kind-icon">Type alias</span></li></ul><ul class="tsd-legend"><li class="tsd-kind-class"><span class="tsd-kind-icon">Class</span></li></ul></div><h2>Settings</h2><p>Theme <select id="theme"><option value="os">OS</option><option value="light">Light</option><option value="dark">Dark</option></select></p></div></footer><div class="container tsd-generator"><p>Generated using <a href="https://typedoc.org/" target="_blank">TypeDoc</a></p></div><div class="overlay"></div><script src="../assets/main.js"></script></body></html>