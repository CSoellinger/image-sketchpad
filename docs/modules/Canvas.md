[image-sketchpad - v1.0.4](../index.md) / [Modules](../modules.md) / Canvas

# Module: Canvas

## Table of contents

### Classes

- [Canvas](../classes/Canvas.Canvas-1.md)

### Type aliases

- [Point](Canvas.md#point)
- [Stroke](Canvas.md#stroke)

## Type aliases

### Point

Ƭ **Point**: `Object`

Point - A point on the canvas set by coordinates.

#### Type declaration

| Name | Type | Description |
| :------ | :------ | :------ |
| `x` | `number` | X coordinate of point. |
| `y` | `number` | Y coordinate of point. |

#### Defined in

[Canvas/Draw.ts:4](https://github.com/CSoellinger/image-sketchpad/blob/main/src/Canvas/Draw.ts#L4)

___

### Stroke

Ƭ **Stroke**: `Object`

Stroke - A line between two or more [Points](Canvas.md#point).

#### Type declaration

| Name | Type | Description |
| :------ | :------ | :------ |
| `cap` | `CanvasLineCap` | {@link CanvasLineCap|Canvas line cap} of the stroke: "butt" | "round" | "square". |
| `color` | `string` | Color of the stroke. |
| `join` | `CanvasLineJoin` | {@link CanvasLineJoin|Canvas line join} of the stroke: "bevel" | "miter" | "round". |
| `maxWidth` | `number` | Max width of the stroke. |
| `miterLimit` | `number` | Miter limit of the stroke. |
| `points` | [`Point`](Canvas.md#point)[] | [Points](Canvas.md#point) of stroke. |
| `width` | `number` | Width of the stroke. |

#### Defined in

[Canvas/Draw.ts:19](https://github.com/CSoellinger/image-sketchpad/blob/main/src/Canvas/Draw.ts#L19)
