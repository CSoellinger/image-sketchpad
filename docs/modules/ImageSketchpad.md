[image-sketchpad - v1.0.4](../index.md) / [Modules](../modules.md) / ImageSketchpad

# Module: ImageSketchpad

## Table of contents

### Classes

- [ImageSketchpad](../classes/ImageSketchpad.ImageSketchpad-1.md)

### Type aliases

- [Options](ImageSketchpad.md#options)
- [UserOptions](ImageSketchpad.md#useroptions)

### Variables

- [DefaultOptions](ImageSketchpad.md#defaultoptions)

## Type aliases

### <a id="options" name="options"></a> Options

Ƭ **Options**: `Object`

Image sketchpad options.

#### Type declaration

| Name | Type | Description |
| :------ | :------ | :------ |
| `enabled` | `boolean` | Enable or disable sketchpad on init.  **`defaultvalue`** `true` |
| `lineCap` | `CanvasLineCap` | Line cap.  **`defaultvalue`** `round` |
| `lineColor` | `string` | Line color.  **`defaultvalue`** `#000` |
| `lineJoin` | `CanvasLineJoin` | Line join.  **`defaultvalue`** `round` |
| `lineMaxWidth` | `number` | Line max width.  **`defaultvalue`** `-1` |
| `lineMiterLimit` | `number` | Line miter limit.  **`defaultvalue`** `10` |
| `lineWidth` | `number` | Line width.  **`defaultvalue`** `5` |
| `writeJsonToHtmlAttribute` | `boolean` | Setting this to true(default) will add a data attribute with the sketchpad json data. This is the same string like you would get with [ImageSketchpad.toJson()](../classes/ImageSketchpad.ImageSketchpad-1.md#tojson).  **`defaultvalue`** `true` |

#### Defined in

[ImageSketchpad/Options.ts:4](https://github.com/CSoellinger/image-sketchpad/blob/main/src/ImageSketchpad/Options.ts#L4)

___

### <a id="useroptions" name="useroptions"></a> UserOptions

Ƭ **UserOptions**: `Partial`<[`Options`](ImageSketchpad.md#options)\>

Image sketchpad user options setting all properties as not required (Partial).

#### Defined in

[ImageSketchpad/Options.ts:66](https://github.com/CSoellinger/image-sketchpad/blob/main/src/ImageSketchpad/Options.ts#L66)

## Variables

### <a id="defaultoptions" name="defaultoptions"></a> DefaultOptions

• `Const` **DefaultOptions**: `Readonly`<[`Options`](ImageSketchpad.md#options)\>

Image sketchpad default options.

#### Defined in

[ImageSketchpad/Options.ts:71](https://github.com/CSoellinger/image-sketchpad/blob/main/src/ImageSketchpad/Options.ts#L71)
