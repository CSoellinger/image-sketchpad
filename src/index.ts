// import jQuery from 'jquery';

import { ImageSketchpad, UserOptions } from './ImageSketchpad';

/**
 * Type definition for instances object
 */
type Instances = Record<string, ImageSketchpad>;

/**
 * Save sketchpad instances to this object
 */
const instances: Instances = {};

/**
 * Initialize new sketchpad or return an already initialized.
 *
 * @param element - HTML image element
 * @param options - Image sketchpad options
 */
const init = (element: HTMLImageElement, options?: UserOptions): ImageSketchpad => {
  if (element?.dataset?.sketchpad && instances[element.dataset.sketchpad]) {
    return instances[element.dataset.sketchpad];
  }

  const newInstance = new ImageSketchpad(element, options);

  if (element?.dataset?.sketchpad) {
    instances[element.dataset.sketchpad] = newInstance;
  }

  return newInstance;
};

// Maybe for future
//
// if (typeof jQuery !== 'undefined') {
//   (function ($: JQueryStatic) {
//     /**
//      * If global variable "jQuery" is defined we will add a new function
//      * called "$().imageSketchpad"
//      *
//      * @param options - Image sketchpad user options
//      */
//     $.fn.imageSketchpad = function (
//       options?: UserOptions
//     ): JQuery {
//       this.each((_index, element: HTMLElement) => {
//         ImageSketchpad(element as HTMLImageElement, options);
//       });

//       return this;
//     };
//   })(jQuery);
// }

export default init;
