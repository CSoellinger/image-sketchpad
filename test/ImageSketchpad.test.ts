import mergeImages from 'merge-images';

import { Canvas } from '../src/Canvas';
import { DefaultOptions, ImageSketchpad } from '../src/ImageSketchpad';

const mockCanvasAdjustFromElement = jest.fn();

jest.mock('merge-images');

jest.mock('../src/Canvas', () => {
  return {
    Canvas: jest.fn().mockImplementation(() => {
      const element = <HTMLCanvasElement>global.document.createElement('canvas');
      const context = element.getContext('2d');

      return {
        element,
        context,
        insert: jest.fn().mockReturnValue(Promise.resolve()),
        adjustFromElement: mockCanvasAdjustFromElement,
        drawStroke: jest.fn().mockReturnValue(Promise.resolve()),
        clear: jest.fn().mockReturnThis(),
      };
    }),
  };
});

describe('Testing ImageSketchpad Class', () => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  let defaultObj: any;
  let image: HTMLImageElement;
  let imageSketchpad: ImageSketchpad | null = null;

  beforeEach(() => {
    document.body.innerHTML = `
      <div>
        <img id="Image" src="../media/fish.png" id="Image" width="800" height="530">
        <div></div>
      </div>`.trim();

    defaultObj = { strokes: [], options: DefaultOptions };
    image = <HTMLImageElement>document.getElementById('Image');
    imageSketchpad = null;

    jest.spyOn(image, 'clientWidth', 'get').mockImplementation(() => 800);
    jest.spyOn(image, 'clientHeight', 'get').mockImplementation(() => 530);
    jest.spyOn(image, 'offsetLeft', 'get').mockImplementation(() => 8);
    jest.spyOn(image, 'offsetTop', 'get').mockImplementation(() => 8);
  });

  afterEach(() => {
    defaultObj.options.lineWidth = 5;
    defaultObj.strokes = [];
    image.dataset['sketchpadJson'] = '';
    document.body.innerHTML = '';
    imageSketchpad?.destroy();
  });

  it('Should initialize and call canvas class constructor one time', () => {
    imageSketchpad = new ImageSketchpad(image, {});

    expect(imageSketchpad).toBeTruthy();
    expect(Canvas).toHaveBeenCalledTimes(1);
    expect(image.classList.contains('sketchpad-loaded')).toBeTruthy();
  });

  it('Should return sketchpad default data as json', () => {
    imageSketchpad = new ImageSketchpad(image);

    return imageSketchpad.toJsonAsync().then((jsonString) => {
      expect(jsonString).toMatchSnapshot();
    });
  });

  it('Should run image load listener with canvas width equals image width', () => {
    image.width = 300;
    imageSketchpad = new ImageSketchpad(image);
    image.dispatchEvent(new Event('load'));

    expect(imageSketchpad).toBeTruthy();
  });

  it('Should run image load listener', () => {
    imageSketchpad = new ImageSketchpad(image);
    image.dispatchEvent(new Event('load'));

    expect(imageSketchpad).toBeTruthy();
  });

  it('Should not run image load listener', () => {
    jest.spyOn(image, 'complete', 'get').mockImplementation(() => true);

    imageSketchpad = new ImageSketchpad(image);
    image.dispatchEvent(new Event('load'));

    expect(imageSketchpad).toBeTruthy();
  });

  it('Should disable and enable sketchpad', () => {
    imageSketchpad = new ImageSketchpad(image);

    expect(imageSketchpad['enabled']).toBeTruthy();

    imageSketchpad.disable();

    expect(imageSketchpad['enabled']).toBeFalsy();

    imageSketchpad.enable();

    expect(imageSketchpad['enabled']).toBeTruthy();
  });

  it('Should trigger canvas adjust from element on window resize', () => {
    imageSketchpad = new ImageSketchpad(image);
    window.dispatchEvent(new Event('resize'));

    expect(mockCanvasAdjustFromElement).toHaveBeenCalledTimes(1);

    window.dispatchEvent(new Event('resize'));

    expect(mockCanvasAdjustFromElement).toHaveBeenCalledTimes(2);

    imageSketchpad.canvas.element.width = 800;
    window.dispatchEvent(new Event('resize'));

    expect(mockCanvasAdjustFromElement).toHaveBeenCalledTimes(2);
  });

  it('Should add and clear a stroke', () => {
    imageSketchpad = new ImageSketchpad(image);

    expect(imageSketchpad['strokes']).toHaveLength(0);

    imageSketchpad['strokes'].push(imageSketchpad['createStroke']([]));

    expect(imageSketchpad['strokes']).toHaveLength(1);

    imageSketchpad.clear();

    expect(imageSketchpad['strokes']).toHaveLength(0);
  });

  it('Should undo and redo an added stroke', () => {
    imageSketchpad = new ImageSketchpad(image);
    const stroke = imageSketchpad['createStroke']([]);
    imageSketchpad['strokes'].push(stroke);

    expect(imageSketchpad['strokes']).toHaveLength(1);
    expect(imageSketchpad['undoneStrokes']).toHaveLength(0);

    imageSketchpad.undo();
    imageSketchpad.undo();

    expect(imageSketchpad['strokes']).toHaveLength(0);
    expect(imageSketchpad['undoneStrokes']).toHaveLength(1);

    imageSketchpad.redo();
    imageSketchpad.redo();

    expect(imageSketchpad['strokes']).toHaveLength(1);
    expect(imageSketchpad['undoneStrokes']).toHaveLength(0);
    expect(imageSketchpad['strokes'][0]).toBe(stroke);
  });

  it('Should add a point to a stroke', () => {
    imageSketchpad = new ImageSketchpad(image);
    const stroke = imageSketchpad['createStroke']([]);
    const point = { x: 8, y: 8 };

    imageSketchpad['strokes'].push(stroke);
    imageSketchpad['pushPoint'](point, stroke);
    imageSketchpad['pushPoint'](point);

    expect(imageSketchpad['strokes'][0]['points']).toHaveLength(2);
  });

  it('Gets a point from event', () => {
    imageSketchpad = new ImageSketchpad(image);
    const mouseDown = new MouseEvent('mousedown', { clientX: 0, clientY: 0 });
    const touchDown = new TouchEvent('touchstart', {
      touches: [<Touch>{ identifier: 1, target: new EventTarget(), pageX: 0, pageY: 0 }],
    });

    expect(imageSketchpad['getPointFromCursor'](mouseDown)).toMatchSnapshot();
    expect(imageSketchpad['getPointFromCursor'](touchDown)).toMatchSnapshot();
  });

  it('Should run listeners', () => {
    const mouseDown = new MouseEvent('mousedown', { clientX: 0, clientY: 0 });
    const mouseMove = new MouseEvent('mousemove', { clientX: 0, clientY: 0 });
    const mouseUp = new MouseEvent('mouseup', { clientX: 0, clientY: 0 });

    const touch = <Touch>{ identifier: 1, target: new EventTarget(), pageX: 0, pageY: 0 };
    const touchDown = new TouchEvent('touchstart', { touches: [touch] });
    const touchMove = new TouchEvent('touchmove', { touches: [touch] });
    const touchEnd = new TouchEvent('touchend', { touches: [touch] });

    imageSketchpad = new ImageSketchpad(image);
    imageSketchpad.canvas.element.dispatchEvent(mouseDown);
    imageSketchpad.canvas.element.dispatchEvent(mouseMove);
    imageSketchpad.canvas.element.dispatchEvent(mouseUp);
    imageSketchpad.canvas.element.dispatchEvent(touchDown);
    imageSketchpad.canvas.element.dispatchEvent(touchMove);
    imageSketchpad.canvas.element.dispatchEvent(touchEnd);

    expect(imageSketchpad['strokes']).toHaveLength(2);

    imageSketchpad.disable();
    imageSketchpad.canvas.element.dispatchEvent(mouseDown);
    imageSketchpad.canvas.element.dispatchEvent(mouseMove);
    imageSketchpad.canvas.element.dispatchEvent(mouseUp);

    expect(imageSketchpad['strokes']).toHaveLength(2);
  });

  it('Should load data sketchpad json attribute', () => {
    defaultObj.options.lineWidth = 7;
    defaultObj.strokes = undefined;

    image.dataset['sketchpadJson'] = JSON.stringify(defaultObj);
    imageSketchpad = new ImageSketchpad(image);

    defaultObj.strokes = [];

    return imageSketchpad.toJsonAsync().then((jsonString) => {
      expect(jsonString).toMatchSnapshot();
    });
  });

  it('Should load data sketchpad options json attribute', () => {
    defaultObj.options.lineWidth = 7;

    image.dataset['sketchpadJson'] = JSON.stringify(defaultObj);
    imageSketchpad = new ImageSketchpad(image);

    return imageSketchpad.toJsonAsync().then((jsonString) => {
      expect(jsonString).toMatchSnapshot();
    });
  });

  it('Should merge the sketch with the image', async () => {
    const base64String =
      'iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mP8z8BQDwAEhQGAhKmMIQAAAABJRU5ErkJggg==';
    (<jest.Mock>mergeImages).mockResolvedValue(base64String);

    imageSketchpad = new ImageSketchpad(image);

    return imageSketchpad.mergeImageWithSketch().then((imageB64) => {
      expect(imageB64).toBe(base64String);
    });
  });

  it('Should download the sketch with the image', () => {
    const base64String =
      'iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mP8z8BQDwAEhQGAhKmMIQAAAABJRU5ErkJggg==';
    (<jest.Mock>mergeImages).mockResolvedValue(base64String);

    imageSketchpad = new ImageSketchpad(image);

    return imageSketchpad.download().then((imageB64) => {
      expect(imageB64).toBe(base64String);
    });
  });

  it('Should download the sketch with the base64 image', () => {
    const base64String =
      'iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mP8z8BQDwAEhQGAhKmMIQAAAABJRU5ErkJggg==';
    (<jest.Mock>mergeImages).mockResolvedValue(base64String);
    image.src = 'data:image/png;base64,' + base64String;

    imageSketchpad = new ImageSketchpad(image);

    return imageSketchpad.download().then((imageB64) => {
      expect(imageB64).toBe(base64String);
    });
  });

  it('Start, draw, stop stroke by mouse event', () => {
    imageSketchpad = new ImageSketchpad(image);

    imageSketchpad['startStrokeHandler'](new MouseEvent('mousedown', { clientX: 0, clientY: 0 }));

    expect(imageSketchpad['sketching']).toBeTruthy();
    expect(imageSketchpad['activeStroke']).toBe(imageSketchpad['strokes'][0]);
    expect(imageSketchpad['strokes']).toHaveLength(1);
    expect(imageSketchpad['strokes'][0]['points']).toHaveLength(1);

    imageSketchpad['drawStrokeHandler'](new MouseEvent('mousemove', { clientX: 0, clientY: 0 }));

    expect(imageSketchpad['strokes'][0]['points']).toHaveLength(2);

    imageSketchpad['endStrokeHandler'](new MouseEvent('mouseup', { clientX: 0, clientY: 0 }));

    expect(imageSketchpad['strokes'][0]['points']).toHaveLength(3);
    expect(image.dataset['sketchpadJson']).toMatchSnapshot();

    imageSketchpad['drawStrokeHandler'](new MouseEvent('mousemove', { clientX: 0, clientY: 0 }));
    imageSketchpad['endStrokeHandler'](new MouseEvent('mouseup', { clientX: 0, clientY: 0 }));

    expect(imageSketchpad['strokes'][0]['points']).toHaveLength(3);
  });

  it('Start, draw, stop stroke by touch event', () => {
    imageSketchpad = new ImageSketchpad(image);

    const touch = <Touch>{ identifier: 1, target: new EventTarget(), pageX: 0, pageY: 0 };
    const touchDown = new TouchEvent('touchstart', { touches: [touch] });
    const touchMove = new TouchEvent('touchmove', { touches: [touch] });
    const touchEnd = new TouchEvent('touchend', { touches: [touch] });

    imageSketchpad['startStrokeHandler'](touchDown);
    imageSketchpad['drawStrokeHandler'](touchMove);
    imageSketchpad['endStrokeHandler'](touchEnd);

    expect(imageSketchpad['strokes'][0]['points']).toHaveLength(2);
    expect(image.dataset['sketchpadJson']).toMatchSnapshot();
  });

  it('Should throw an error if we try loading a bad JSON string', () => {
    imageSketchpad = new ImageSketchpad(image, {});

    return imageSketchpad.loadJson('BAD_JSON').catch((err) => {
      // eslint-disable-next-line jest/no-conditional-expect
      expect(err).toBeInstanceOf(Error);
    });
  });

  it('Should throw an error we load it on non image html elements', () => {
    expect(() => {
      imageSketchpad = new ImageSketchpad(<HTMLImageElement>document.createElement('div'));
    }).toThrow();

    expect(imageSketchpad).toBeNull();
  });

  it('Should throw an error if we try double initialization', () => {
    imageSketchpad = new ImageSketchpad(image);
    imageSketchpad = null;

    expect(() => {
      imageSketchpad = new ImageSketchpad(image);
    }).toThrow();

    expect(imageSketchpad).toBeNull();
  });
});
