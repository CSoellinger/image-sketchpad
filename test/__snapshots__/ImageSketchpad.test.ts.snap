// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing ImageSketchpad Class Gets a point from event 1`] = `
Object {
  "x": 0,
  "y": 0,
}
`;

exports[`Testing ImageSketchpad Class Gets a point from event 2`] = `
Object {
  "x": 0,
  "y": 0,
}
`;

exports[`Testing ImageSketchpad Class Should load data sketchpad json attribute 1`] = `"{\\"strokes\\":[],\\"options\\":{\\"lineWidth\\":7,\\"lineMaxWidth\\":-1,\\"lineColor\\":\\"#000\\",\\"lineCap\\":\\"round\\",\\"lineJoin\\":\\"round\\",\\"lineMiterLimit\\":10}}"`;

exports[`Testing ImageSketchpad Class Should load data sketchpad options json attribute 1`] = `"{\\"strokes\\":[],\\"options\\":{\\"lineWidth\\":7,\\"lineMaxWidth\\":-1,\\"lineColor\\":\\"#000\\",\\"lineCap\\":\\"round\\",\\"lineJoin\\":\\"round\\",\\"lineMiterLimit\\":10}}"`;

exports[`Testing ImageSketchpad Class Start, draw, stop stroke by mouse event 1`] = `"{\\"strokes\\":[{\\"points\\":[{\\"x\\":0,\\"y\\":0},{\\"x\\":0,\\"y\\":0}],\\"width\\":5,\\"maxWidth\\":-1,\\"color\\":\\"#000\\",\\"cap\\":\\"round\\",\\"join\\":\\"round\\",\\"miterLimit\\":10}],\\"options\\":{\\"lineWidth\\":5,\\"lineMaxWidth\\":-1,\\"lineColor\\":\\"#000\\",\\"lineCap\\":\\"round\\",\\"lineJoin\\":\\"round\\",\\"lineMiterLimit\\":10}}"`;

exports[`Testing ImageSketchpad Class Start, draw, stop stroke by touch event 1`] = `"{\\"strokes\\":[{\\"points\\":[{\\"x\\":0,\\"y\\":0},{\\"x\\":0,\\"y\\":0}],\\"width\\":5,\\"maxWidth\\":-1,\\"color\\":\\"#000\\",\\"cap\\":\\"round\\",\\"join\\":\\"round\\",\\"miterLimit\\":10}],\\"options\\":{\\"lineWidth\\":5,\\"lineMaxWidth\\":-1,\\"lineColor\\":\\"#000\\",\\"lineCap\\":\\"round\\",\\"lineJoin\\":\\"round\\",\\"lineMiterLimit\\":10}}"`;
